### PR TITLE
change user name component

### PR DESCRIPTION
**Pull request by linusarnejohansson**
not ready yet.

### CODE REVIEW

**Comment by linusarnejohansson** [comment]
the changes are tested to work with:
- changing name
- changing password

**Comment by linusarnejohansson** [comment]
need the changes in https://github.com/userwidgets/model/pull/59 to work

File: src/State/Users.ts [hunk 1]
```diff
@@ -61,20 +61,38 @@ export class Users extends smoothly.StateBase<Users, userwidgets.ClientCollectio
 			this.request = undefined
 		return (this.listenable.value = result) || false
 	}
-	async updatePermissions(
-		email: string,
-		permissions: userwidgets.User.Permissions.Readable
+	async update( //this function requires more control and checks. More error messages. Also a success message if everything was successful.
```
Comment (EliasEriksson) [inline src/State/Users.ts]: remove comment

File: src/State/Users.ts [hunk 2]
```diff
@@ -61,20 +61,38 @@ export class Users extends smoothly.StateBase<Users, userwidgets.ClientCollectio
...
+		options?: { entityTag?: string }
 	): Promise<userwidgets.User | false> {
-		const promise = !this.state.organizations.current
+		const promise = !this.state.me.key
 			? undefined
 			: this.client.user
-					.updatePermissions(email, this.state.organizations.current.id, permissions)
+					.update(email, user, options)
 					.then(response => (!userwidgets.User.is(response) ? false : response))
+		// this should reauthenticate the user by using the login function(?) but giving it the current token instead of the email and password as credentials
```
Comment (EliasEriksson) [inline src/State/Users.ts]: remvoe comment

File: src/State/Users.ts [hunk 3]
```diff
@@ -61,20 +61,38 @@ export class Users extends smoothly.StateBase<Users, userwidgets.ClientCollectio
...
+		// maybe modify login function to handle token as well?
 		const result = await promise
-		if (result)
-			this.fetch()
+		console.log("state update", result, "this.state.me.key", this.state.me.key)
```
Comment (EliasEriksson) [inline src/State/Users.ts:76]: remove log

File: src/State/Users.ts [hunk 4]
```diff
@@ -61,20 +61,38 @@ export class Users extends smoothly.StateBase<Users, userwidgets.ClientCollectio
...
+	// 	const promise = !this.state.organizations.current
+	// 		? undefined
+	// 		: this.client.user
+	// 				.updatePermissions(email, this.state.organizations.current.id, permissions)
+	// 				.then(response => (!userwidgets.User.is(response) ? false : response))
+	// 	const result = await promise
+	// 	if (result)
+	// 		this.fetch()
+	// 	return result || false
+	// }
```
Comment (EliasEriksson) [inline src/State/Users.ts]: this section can be removed

File: src/components/me/name/index.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+		console.log("inputHandler, this.change", this.change, "event.detail", event.detail)
+	}
+	async submitHandler(event: CustomEvent<smoothly.Data>) {
+		this.inputHandler(event)
+		const name = this.change?.name
+		if (!userwidgets.User.Name.is(name))
+			this.notice.emit(smoothly.Notice.failed("Malformed name."))
+		else if (!this.token)
+			this.notice.emit(smoothly.Notice.failed("need a token"))
+		//fix error message and flow, needs more checks(?)
```
Comment (EliasEriksson) [inline src/components/me/name/index.tsx]: remove

File: src/components/me/name/index.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+					<smoothly-input
+						readonly={!this.change}
+						name="name.last"
+						value={this.user ? this.user.name.last : this.token ? this.token.name.last : null}>
+						Last name
+					</smoothly-input>
+					<userwidgets-edit-button
+						slot="submit"
+						state={this.state}
+						disabled={!!this.request} //add check to disable if this.change.name == original name
```
Comment (EliasEriksson) [inline src/components/me/name/index.tsx]: remove comment

File: src/components/me/name/index.tsx [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+			<Host>
+				<smoothly-form
+					looks="border"
+					onSmoothlyFormInput={e => this.inputHandler(e)}
+					onSmoothlyFormSubmit={e => this.submitHandler(e)}>
+					<smoothly-input
+						readonly={!this.change}
+						name="name.first"
+						value={this.user ? this.user.name.first : this.token ? this.token.name.first : null}>
+						First name
```
Comment (EliasEriksson) [inline src/components/me/name/index.tsx:71]: add translation

File: src/components/me/name/index.tsx [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+					</smoothly-input>
+					<smoothly-input
+						readonly={!this.change}
+						name="name.last"
+						value={this.user ? this.user.name.last : this.token ? this.token.name.last : null}>
+						Last name
```
Comment (EliasEriksson) [inline src/components/me/name/index.tsx:77]: add translation

File: src/components/me/name/index.tsx [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+				...this.change,
+				name: { ...this.change.name, ...(typeof event.detail.name == "object" && event.detail.name) },
+			})
+		console.log("inputHandler, this.change", this.change, "event.detail", event.detail)
+	}
+	async submitHandler(event: CustomEvent<smoothly.Data>) {
+		this.inputHandler(event)
+		const name = this.change?.name
+		if (!userwidgets.User.Name.is(name))
+			this.notice.emit(smoothly.Notice.failed("Malformed name."))
```
Comment (EliasEriksson) [inline src/components/me/name/index.tsx:49]: add translation

File: src/components/me/name/index.tsx [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+		else if (!this.token)
+			this.notice.emit(smoothly.Notice.failed("need a token"))
```
Comment (EliasEriksson) [inline src/components/me/name/index.tsx:51]: add translation

File: src/components/me/name/index.tsx [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+		//fix error message and flow, needs more checks(?)
+		else if (!(await this.state.users.update(this.user?.email ?? this.token.email, { name })))
+			this.notice.emit(smoothly.Notice.failed("failed to update name"))
```
Comment (EliasEriksson) [inline src/components/me/name/index.tsx:53]: add translation

File: src/components/me/name/index.tsx [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+		else {
+			this.notice.emit(smoothly.Notice.succeeded("Your name has been updated"))
```
Comment (EliasEriksson) [inline src/components/me/name/index.tsx:55]: add translation

**Review by EliasEriksson — CHANGES_REQUESTED** [review]
some fixes

**Review by EliasEriksson — APPROVED** [review]

**Pull request merged**